common/src/main/scala/explore/model/ConstraintsUndoStacks.scala

```diff
@@ -0,0 +1,45 @@
+// Copyright (c) 2016-2021 Association of Universities for Research in Astronomy, Inc. (AURA)
+// For license information see LICENSE or https://opensource.org/licenses/BSD-3-Clause
+
+package explore.model
+
+import cats.Eq
+import cats.syntax.all._
+import explore.common.ConstraintGroupQueries.ConstraintGroupList
+import explore.undo.UndoStacks
+import lucuma.core.model.Observation
+
+import scala.collection.immutable.SortedSet
+
+sealed trait ConstraintsUndoStacks[F[_]] {
+  import ConstraintsUndoStacks._
+
+  def listStacks: UndoStacks[F, ConstraintGroupList] = this match {
+    case ConstraintGroupUndoStacks(_, _) => UndoStacks.empty[F, ConstraintGroupList]
+    case ConstraintListUndoStacks(us)    => us
+  }
+
+  def groupStacks(obsIds: SortedSet[Observation.Id]): UndoStacks[F, ConstraintSet] = this match {
+    case ConstraintGroupUndoStacks(ids, undoStacks) if ids === obsIds => undoStacks
+    case _                                                            => UndoStacks.empty[F, ConstraintSet]
+  }
+}
+
+object ConstraintsUndoStacks {
+  case class ConstraintGroupUndoStacks[F[_]](
+    ids:        SortedSet[Observation.Id],
```

Comment:
This is a question that goes beyond the scope of this PR but I've been curious for a while: why is this a `SortedSet` and not just a `Set`?

Reply:
At least in the past I often used `SortedSet/Map` in places because monocle would provide some lenses only for sorted collections. This was relaxed on the latest versions of monocle

Reply:
I'm sorting the constraint group list by lowest observation id, and this seemed the easiest way to accomplish that. (Although it turned out that the `Order` defined by cats for SortedSets didn't work for that purpose, so I had to define a new one. So, maybe there is a better way to do it.)

I chose that sort order because it seemed the most stable ordering for the list. There isn't a constraint id, and an ordering by name would potentially be reordering the list with every edit. It seems a bit arbitrary, but I figured I'd do it that way until I got feedback requesting otherwise.

Reply:
Ah I see now. Having the `Observation.Id`s sorted within each group is an optimization when sorting the groups.